### PR TITLE
Don't merge CGUs in incremental builds.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
The idea here is that the default number of CGUs for incremental builds is 256. This is a high number. Most benchmarks don't have that many CGUs. Only the biggest benchmarks do, which means they tend to do worse with `incr-patched` results, because CGU merging generally interferes with partial incremental rebuilds. So why not let those larger benchmarks have more CGUs?

The drastic results for incremental opt builds of `hyper` are because it sets `codegen-units = 1` for release builds, which means incremental normally has almost no effect there, and this PR effectively overrides that. So those can be mostly ignored. (There's an argument that we shouldn't measure opt incremental builds in rustc-perf, because it's not a default configuration and probably very rarely used in the real world.)

The big swings for `bitmaps` are because it has become very noisy :(

The -4.17% walltime for `cargo-0.60.0 debug incr-patched: println` is the kind of thing I was expecting to see -- a benchmark with more than 256 CGUs seeing an improvement for an `incr-patched` run due to less CGU merging. But I thought some of the other big cases (such as `webrender`, `cranelift-codegen`, and `ripgrep`) would also see improvements.

A good comparison is [this result](https://perf.rust-lang.org/compare.html?start=ad23942ad4079f779651983b262fa34c0beda701&end=26269dfb9e9b933270bec418381c5dbf44934013&stat=wall-time) in #111666 where I reduced the number of CGUs in incremental builds from 256 to 64, and there were some drastic `incr-patched` regressions on the larger benchmarks. That's the kind of regression this PR's changes would avoid. The `incr-patched` runs here are pretty basic and don't give much coverage, it's easy to imagine that they're just getting lucky, and that changes to other parts of the code could have larger effects with only 256 CGUs.

File: compiler/rustc_monomorphize/src/partitioning.rs [hunk 1]
```diff
@@ -161,10 +161,11 @@ where
 
     debug_dump(tcx, "INITIAL PARTITIONING", &codegen_units);
 
+    // njn: update comment
     // Merge until we have at most `max_cgu_count` codegen units.
     // `merge_codegen_units` is responsible for updating the CGU size
     // estimates.
-    {
+    if cx.tcx.sess.opts.incremental.is_none() {
```
Comment (bjorn3) [inline compiler/rustc_monomorphize/src/partitioning.rs:168]: Currently `-Ccodegen-units` overrides the 256 default for incr comp. With this change that no longer happens. I think user's intent should be preserved even with incr comp. Maybe they care about runtime perf too and many codegen units is a huge regression (eg in a game) or maybe duplication of inlined functions costs more than not merging saves in reusable cgus. Either way the user should have the option to do what is best for them.

Comment (nnethercote) [reply compiler/rustc_monomorphize/src/partitioning.rs:168]: Sure, as written I've done this too crudely. If this were to land various other changes would be needed, e.g. documentation. I have changed the PR to draft status to better reflect this.

**Comment by michaelwoerister** [comment]
To second @bjorn3's comment: I think we should keep having a way for controlling the max number of CGUs for incremental builds. Making incr. comp. respect `-Ccodegen-units` was introduced as a feature at some point because it allows to keep resource usage in check (there's quite a bit of fixed per-CGU overhead, IIRC).

I'm not against doing experiments in this area, of course. I just want to make sure this is taking into account.

**Comment by nnethercote** [comment]
Respecting `-Ccodegen-units` is definitely a good idea. But do people think that defaulting to an unlimited number of CGUs is reasonable?

**Comment by michaelwoerister** [comment]
I'm not sure about unlimited[^1] but I have no objection to increasing the default to a much higher number. If I remember correctly, the current default was introduced in the same PR that made incr. comp. take `-Ccodegen-units` into account and was basically calibrated towards the then current set of perf.rlo benchmarks. From my point of view, the important thing is that users have the ability to control the number of CGUs if the default doesn't work for them. 


[^1]: The initial set of CGUs is still 2 CGUs for each source-level module, right? I could imagine that some macros or other kind of code generation will create lots of small modules, so I'd be a bit worried about not setting an upper bound at all. The bound could be rather high (1000? 2000?) -- but I'd be worried about users discovering that their incr. cache directory is a 100Gb and contains 50000 object files and not quite knowing what to do about that. I think that would be worse than less effective incr. comp. for extremely large crates.

**Comment by nnethercote** [comment]
Sounds like the simplest path forward would be to lift the limit from 256 CGUs to perhaps 512 or 1024. But now I'm not sure it's worth the effort.

**Pull request closed without merging**